### ISSUE TITLE
[Bug] Eagle3 overlap scheduling decreases performance

### ISSUE THREAD

**Issue by TJ5** [sgl-project/sglang#12411]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

I observe decreased performance when using overlap scheduling with eagle3 speculative decoding (enabled by passing environment variable SGLANG_ENABLE_SPEC_V2=1). Specifically inference speed was significantly lower even on low concurrency. 

### Reproduction

From https://github.com/sgl-project/sglang/pull/11821, overlap scheduling is supported for eagle with triton backend. Benchmarked triton with no spec decoding and compared with the new feature.

We use [openai/gpt-oss-120b](https://huggingface.co/openai/gpt-oss-120b) model with [nvidia/gpt-oss-120b-Eagle3](https://huggingface.co/nvidia/gpt-oss-120b-Eagle3) draft model. 

Baseline:
```
python3 -m sglang.launch_server \
    --model-path openai/gpt-oss-120b \
    --tp 2 \
    --port 8080 \
    --attention-backend=triton \
    --host 0.0.0.0 \
    --log-requests-level=1 \
    --tool-call-parser=gpt-oss \
    --reasoning-parser=gpt-oss
```
EAGLE3 Server command:
```
SGLANG_ENABLE_SPEC_V2=1 python3 -m sglang.launch_server \
    --model-path /raid/models/openai/gpt-oss-120b \
    --tp 2 \
    --port 8080 \
    --attention-backend=triton \
    --host 0.0.0.0 \
    --log-requests-level=1 \
   --tool-call-parser=gpt-oss \
--reasoning-parser=gpt-oss \
--speculative-draft-model-path /raid/models/nvidia/gpt-oss-120b-Eagle3 \
--speculative-algorithm EAGLE3
```

Benchmarked performance with [genai-bench](https://github.com/sgl-project/genai-bench), saw the following results: 

<img width="3200" height="1200" alt="Image" src="https://github.com/user-attachments/assets/17f766bd-9f3c-44da-bb0f-04b498943df7" />

Throughput is expected to be lower, but we would expect that output inference speed (rate of token generation on a single request) should be higher with speculative decoding than without. However, even when concurrency=1, the overlap scheduler with eagle3 is slower than without any speculative decoding. 

The benchmark was run with the [ZAAAAAC/qwen2.5_test_python](https://huggingface.co/datasets/ZAAAAAC/qwen2.5_test_python) dataset and can be reproduced with: 

Benchmark baseline: 
```
genai-bench benchmark \
--api-backend openai \
--api-key "..." \
--api-base "$BASELINE_URL" \
--api-model-name "sglang-model" \
--model-tokenizer "/raid/models/openai/gpt-oss-120b" \
--task text-to-text \
--server-engine SGLang \
--server-gpu-type H100 \
--server-version dev \
--server-gpu-count 2  \
--max-time-per-run 10 \
--max-requests-per-run 300 \
--experiment-base-dir $RESULTS_DIR \
--num-concurrency 1 \
--num-concurrency 4 \
--num-concurrency 16 \
--num-concurrency 64 \
--num-concurrency 256 \
--traffic-scenario dataset \
--dataset-config $DATASET_CONFIG
```
Benchmark overlap scheduled eagle3
```
genai-bench benchmark \
--api-backend openai \
--api-key "..." \
--api-base "$EAGLE3_URL" \
--api-model-name "sglang-model" \
--model-tokenizer "/raid/models/openai/gpt-oss-120b" \
--task text-to-text \
--server-engine SGLang \
--server-gpu-type H100 \
--server-version dev \
--server-gpu-count 2  \
--max-time-per-run 10 \
--max-requests-per-run 300 \
--experiment-base-dir $RESULTS_DIR \
--num-concurrency 1 \
--num-concurrency 4 \
--num-concurrency 16 \
--num-concurrency 64 \
--num-concurrency 256 \
--traffic-scenario dataset \
--dataset-config $DATASET_CONFIG
```
Where the dataset config is just a json of the form: 
```
{
  "source": {
    "type": "huggingface",
    "path": "ZAAAAAC/qwen2.5_test_python",
    "huggingface_kwargs": {
      "split": "train"
    }
  },
  "prompt_column": "instruction"
}
```

### Environment

Python: 3.13.2 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:02) [GCC 11.2.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H100 80GB HBM3
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr
NVCC: Cuda compilation tools, release 11.5, V11.5.119
CUDA Driver Version: 550.163.01
PyTorch: 2.8.0+cu128
sglang: 0.5.4.post1
sgl_kernel: 0.3.16.post4
flashinfer_python: 0.4.1
triton: 3.4.0
transformers: 4.57.1
torchao: 0.9.0
numpy: 2.3.4
aiohttp: 3.13.2
fastapi: 0.120.2
hf_transfer: 0.1.9
huggingface_hub: 0.36.0
interegular: 0.3.3
modelscope: 1.31.0
orjson: 3.11.4
outlines: 0.1.11
packaging: 25.0
psutil: 7.1.2
pydantic: 2.12.3
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.38.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.1.25
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.72.0
litellm: Module Not Found
decord2: 2.0.0
NVIDIA Topology: 
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	NIC2	NIC3	NIC4	NIC5	NIC6	NIC7	NIC8	NIC9	NIC10	NIC11	NIC12	NIC13	NIC14	NIC15	NIC16NIC17	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NV18	PXB	PXB	NODE	NODE	NODE	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	0-55,112-167	0		N/A
GPU1	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NODE	NODE	NODE	PXB	PXB	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	0-55,112-167	0		N/A
GPU2	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NODE	NODE	NODE	NODE	NODE	PXB	PXB	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	0-55,112-167	0		N/A
GPU3	NV18	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NODE	NODE	NODE	NODE	NODE	NODE	NODE	PXB	PXB	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	0-55,112-167	0		N/A
GPU4	NV18	NV18	NV18	NV18	 X 	NV18	NV18	NV18	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	PXB	PXB	NODE	NODE	NODE	NODE	NODE	NODE	NODE	56-111,168-223	1		N/A
GPU5	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NV18	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PXB	PXB	NODE	NODE	NODE	NODE	56-111,168-223	1		N/A
GPU6	NV18	NV18	NV18	NV18	NV18	NV18	 X 	NV18	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	PXB	PXB	NODE	NODE	56-111,168-223	1		N/A
GPU7	NV18	NV18	NV18	NV18	NV18	NV18	NV18	 X 	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	PXB	PXB	56-111,168-223	1		N/A
NIC0	PXB	NODE	NODE	NODE	SYS	SYS	SYS	SYS	 X 	PIX	NODE	NODE	NODE	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC1	PXB	NODE	NODE	NODE	SYS	SYS	SYS	SYS	PIX	 X 	NODE	NODE	NODE	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC2	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	NODE	NODE	 X 	NODE	NODE	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC3	NODE	PXB	NODE	NODE	SYS	SYS	SYS	SYS	NODE	NODE	NODE	 X 	PIX	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC4	NODE	PXB	NODE	NODE	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PIX	 X 	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC5	NODE	NODE	PXB	NODE	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	 X 	PIX	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC6	NODE	NODE	PXB	NODE	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	PIX	 X 	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC7	NODE	NODE	NODE	PXB	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	 X 	PIX	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC8	NODE	NODE	NODE	PXB	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	PIX	 X 	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS				
NIC9	SYS	SYS	SYS	SYS	PXB	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	 X 	PIX	NODE	NODE	NODE	NODE	NODE	NODE	NODE				
NIC10	SYS	SYS	SYS	SYS	PXB	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	PIX	 X 	NODE	NODE	NODE	NODE	NODE	NODE	NODE				
NIC11	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	 X 	NODE	NODE	NODE	NODE	NODE	NODE				
NIC12	SYS	SYS	SYS	SYS	NODE	PXB	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	 X 	PIX	NODE	NODE	NODE	NODE				
NIC13	SYS	SYS	SYS	SYS	NODE	PXB	NODE	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PIX	 X 	NODE	NODE	NODE	NODE				
NIC14	SYS	SYS	SYS	SYS	NODE	NODE	PXB	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	 X 	PIX	NODE	NODE				
NIC15	SYS	SYS	SYS	SYS	NODE	NODE	PXB	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	PIX	 X 	NODE	NODE				
NIC16	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PXB	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	 X 	PIX				
NIC17	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PXB	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	PIX	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5
  NIC6: mlx5_6
  NIC7: mlx5_7
  NIC8: mlx5_8
  NIC9: mlx5_9
  NIC10: mlx5_10
  NIC11: mlx5_11
  NIC12: mlx5_12
  NIC13: mlx5_13
  NIC14: mlx5_14
  NIC15: mlx5_15
  NIC16: mlx5_16
  NIC17: mlx5_17

ulimit soft: 131072
